packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts

```diff
@@ -0,0 +1,124 @@
+import {
+  ConnectionConfig,
+  InteractionContext,
+  InteractionType,
+  createInteractionContext
+} from '@cardano-ogmios/client';
+import { Observable, switchMap } from 'rxjs';
+import { RetryBackoffConfig, retryBackoff } from 'backoff-rxjs';
+import { WithLogger, contextLogger, isConnectionError } from '@cardano-sdk/util';
+
+export type ReconnectionConfig = Omit<RetryBackoffConfig, 'shouldRetry'>;
+
+const defaultReconnectionConfig: ReconnectionConfig = { initialInterval: 10, maxInterval: 5000 };
+
+export interface InteractionContextProps {
+  /**
+   * Subscribed when making an initial connection and
+   * re-subscribed before reconnecting on connection errors.
+   *
+   * Emitting new value from this observable will
+   * close an existing connection and open a new one.
+   */
+  connectionConfig$: Observable<ConnectionConfig>;
+  /**
+   * 'LongRunning' will keep the connection open until unsubscribed.
+   * 'OneTime' will close the connectino (and complete the interaction context observable)
+   * after the 1st usage of emitted interaction context.
+   */
+  interactionType: InteractionType;
+  /**
+   * Retry backoff configuration for
+   * re-subscribing to connectionConfig$ on connection error.
+   */
+  reconnectionConfig?: ReconnectionConfig;
+}
+
+/**
+ * Creates an Ogmios InteractionContext upon subscription.
+ * Closes the connection when unsubscribed.
+ * Reconnects and emits a new InteractionContext upon connection errors.
+ *
+ * @throws errors with UnknownCardanoNodeError on any non-connection
+ * InteractionContext error thrown by Ogmios.
+ * @throws errors with CardanoNodeErrors.ConnectionError when `reconnectionConfig.maxAttempts` is reached.
+ */
+export const createObservableInteractionContext = (
+  { connectionConfig$, interactionType, reconnectionConfig = defaultReconnectionConfig }: InteractionContextProps,
+  dependencies: WithLogger
+) =>
+  connectionConfig$.pipe(
+    switchMap(
+      (connection) =>
+        new Observable<InteractionContext>((subscriber) => {
+          const logger = contextLogger(dependencies.logger, 'InteractionContext');
+          const interactionContextReady = createInteractionContext(
+            (error) => {
+              logger.error(error.message);
+              subscriber.error(
+                error instanceof CardanoNodeErrors.CardanoClientErrors.ConnectionError
+                  ? error
+                  : new CardanoNodeErrors.UnknownCardanoNodeError(error)
+              );
+            },
+            (code, reason) => {
+              if (code === 1000) {
+                logger.debug('Websocket closed with code 1000 (Normal Closure)');
+                subscriber.complete();
+              } else {
+                const message = `Websocket unexpectedly closed with code ${code}: ${reason}`;
+                logger.error(message);
+                // TODO: this error doesn't contain any details {code,reason}, maybe create a new error type?
```

Comment:
Sounds reasonable to define a custom error

Reply:
Agree, but we should be using CardanoNodeErrors in this context, and all of the errors used are currently Ogmios. If we add a new type, then we need to make sure to map ogmios ConnectionError into the new error everywhere else. If we don't then we end up with 2 connection error types that represent the same thing. Maybe let's tackle this separately?

Reply:
Yes, separate is best.

Reply:
ADP-2734
[Updated the comment](https://github.com/input-output-hk/cardano-js-sdk/pull/593/commits/b1c7785c2b7cd554dc1adf9cf90db5897e4eaebc#diff-d37d215964cf45049fba9ad64ec4cbe222235c119e6890b805c781833edff19cR72-R73) in appropriate commit